zilliqa/src/exec.rs

```diff
@@ -1323,13 +1327,56 @@ impl From<Account> for PendingAccount {
     }
 }
 
+fn cache_external_libraries(
+    state: &State,
+    ext_libs_cache_dir: &str,
+    ext_libraries: &[ExternalLibrary],
+) -> Result<()> {
+    let ext_libs_path = Path::new(ext_libs_cache_dir);
+
+    for lib in ext_libraries {
+        let account = state.get_account(lib.address)?;
+        match &account.code {
+            Code::Evm(_) => {
+                return Err(anyhow!(
+                    "Impossible to load an EVM contract as a Scilla library."
+                ));
+            }
+            Code::Scilla {
+                code, init_data, ..
+            } => {
+                let contract_init = ContractInit::new(serde_json::from_str(init_data)?)?;
+                if !contract_init.is_library()? {
+                    return Err(anyhow!(
+                        "Impossible to load a non-library contract as a Scilla library."
+                    ));
+                }
+
+                let file_path = ext_libs_path.join(&lib.name);
+                let canonical_file_path = fs::canonicalize(&file_path)?;
+
+                // To prevent malicious library names like ../../../a.txt
+                if !canonical_file_path.starts_with(&ext_libs_path) {
+                    return Err(anyhow::anyhow!("Invalid library name: {}", lib.name));
+                }
+
+                fs::write(&canonical_file_path, code).with_context(|| {
+                    format!("Failed to write the contract code to {:?}", file_path)
+                })?;
+            }
+        }
+    }
+    Ok(())
+}
+
 fn scilla_create(
     mut state: PendingState,
     scilla: MutexGuard<'_, Scilla>,
     from_addr: Address,
     txn: TxZilliqa,
     current_block: BlockHeader,
     mut inspector: impl ScillaInspector,
+    ext_libs_cache_dir: &str,
```

Comment:
How are external libraries cleaned up after execution?

Reply:
I had to use `scopeguard` crate to ensure they'll be deleted even in case of an error.